File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool dfs(TreeNode* p, TreeNode* q, bool &ans)
    {
        if((p == nullptr && q != nullptr) || (p != nullptr && q == nullptr) || ((p&&q) && (p->val != q->val)) || !ans)
        {
            ans = false;
            return false;
        }        
        if((!p&&!q))
        {
            return true;
        }
        
        
        dfs(p->right,q->right,ans);
        dfs(p->left,q->left,ans);

        return ans;
    }
    
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if(!root && !subRoot)
        {
            return false;
        }
        stack<TreeNode*> st;
        st.push(root);
        
        while(!st.empty())
        {
            bool ans = true;
            TreeNode* temp = st.top();
            st.pop();
            ans = dfs(temp,subRoot,ans);
            cout<<(ans);
            if(ans==true)
            {
                return true;
            }
            if(temp->right) { st.push(temp->right);}
            if(temp->left) {st.push(temp->left);}

        }
        return false;
        
        
        
        
    }
};

File: Data Structures & Algorithms/spiral-matrix/submission-0.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m = len(matrix)
        n = len(matrix[0])
        top = 0
        bottom = m-1
        left = 0
        right = n-1
        ans = []

        while left <= right and top <= bottom:
            #left -> right
            print(left,right)
            
            i = left
            for j in range(left, right+1):
                ans.append(matrix[i][j])

            #right -> bottom
            j = right
            for i in range(top+1, bottom+1):
                ans.append(matrix[i][j])

            #bottom ->left
            i = bottom
            if top < bottom:
                for j in range(right-1, left-1,-1):
                        ans.append(matrix[i][j])

            #left -> top
            j = left
            if left < right:
                for i in range(bottom-1,top,-1):
                    ans.append(matrix[i][j])
            
            top += 1
            bottom -= 1
            left += 1
            right -= 1
        
        return ans






        